pyteal/compiler/optimizer/optimizer.py

```diff
@@ -0,0 +1,89 @@
+from typing import cast, Set
+from pyteal.ir.tealblock import TealBlock
+
+from pyteal.ir.tealop import TealOp
+from pyteal.ir.ops import Op
+
+
+class OptimizeOptions:
+    """An object which specifies the optimizations to be performed and relevant context.
+    Args:
+
+        scratch_slots (optional): cancel contiguous store/load operations
+            that have no load dependencies elsewhere.
+        skip_ids (optional): the slot ids that should be skipped during
+            optimization. At the moment this includes:
+                1. reserved slots because they may have dependencies outside
+                the current application. For example, the 'gloads' opcode can
+                access the slots of other applications in the tx group.
+                2. global slots because they're outside the scope of global
+                optimizations, which only apply to the control flow graph of
+                a single subroutine.
+                3. slots used with dynamic scratch vars. These slots use
+                indirection by means of the 'stores' opcode and dependencies
+                can only be determined at runtime."""
+
+    def __init__(self, *, scratch_slots: bool = False, skip_ids: Set[int] = None):
+        self.scratch_slots = scratch_slots
+        self.skip_ids: Set[int] = skip_ids if skip_ids is not None else set()
+
+
+def _remove_extraneous_slot_access(start: TealBlock, remove: Set[int]):
+    def keep_op(op: TealOp):
+        if type(op) != TealOp or (op.op != Op.store and op.op != Op.load):
+            return True
+
+        return cast(int, op.args[0]) not in remove
+
+    for block in TealBlock.Iterate(start):
+        block.ops = list(filter(keep_op, block.ops))
+
+
+def _has_load_dependencies(cur_block: TealBlock, start: TealBlock, slot: int, pos: int):
+    for block in TealBlock.Iterate(start):
+        for i, op in enumerate(block.ops):
+            if block == cur_block and i == pos:
+                continue
+
+            if type(op) == TealOp and op.op == Op.load and op.args[0] == slot:
+                return True
+
+    return False
+
+
+def _apply_slot_to_stack(cur_block: TealBlock, start: TealBlock, skip_ids: Set[int]):
+    slots_to_remove = set()
+    for i, op in enumerate(cur_block.ops[:-1]):
```

Comment:
It's possible for some blocks to have 0 ops (an example is the program `If(Int(1)).Then(Approve()).Else(Reject())`, since the branch converges on an empty block), and if that's the case I think this would error.

A very simple unit test that an empty block won't cause this to fail would be nice to have.

Reply:
I think the inner for loop in `apply_global_optimizations` prevents calling `_apply_slot_to_stack` with an empty block. Surprisingly, even if an empty block were passed directly to `_apply_slot_to_stack`, it wouldn't cause an error because it's completely legal in python for slice indexes to be outside the bounds of the underlying list. I added tests to confirm this behavior.

Reply:
Interesting, thanks for looking into this!